File: app/src/main/java/k/bs/designpatternsp/pattern/prototype/messagebox/PrototypeService.java
package k.bs.designpatternsp.pattern.prototype.messagebox;

import java.util.HashMap;


public class PrototypeService {
    private HashMap showcase = new HashMap<>();

    public void register(String name, Product proto) {
        showcase.put(name, proto);
    }

    public Product clone(String protoName) {
        return ((Product) showcase.get(protoName)).createClone();
    }
}

File: app/src/main/java/k/bs/designpatternsp/pattern/prototype/bike/PrototypeService.kt
package k.bs.designpatternsp.pattern.prototype.bike

class PrototypeService<T : Vehicle> {
    private val showcase = hashMapOf<String, T>()

    fun register(name: String, model: T) {
        showcase[name] = model
    }

    fun createClone(name: String): T? {
        return (showcase[name]?.createClone()) as T
    }
}